Selfie PR (#7)
igorganich/damn-vulnerable-defi-halmos: test/selfie/README.md

```solidity
function onFlashLoan(address initiator, address token,
                    uint256 amount, uint256 fee,
                    bytes calldata data
) external returns (bytes32) 
{
    address target = svm.createAddress("target_onFlashLoan");
    bytes memory data;
    //Get some concrete target-name pair
    (target, data) = glob.get_concrete_from_symbolic(target);
    target.call(data);
}
```
Plus, take into account that this `flashLoan` needs to be returned "honestly" this time. And should return a valid string:
```solidity
function flashLoan(...)
{
    ...
    if (!token.transferFrom(address(_receiver), address(this), _amount)) {
        revert RepayFailed();
    }
    ...
}
```
So, **SymbolicAttacker**:
```solidity
function onFlashLoan(...)
{
    ...
    DamnValuableVotes(token).approve(address(msg.sender), 2**256 - 1);
    return (keccak256("ERC3156FlashBorrower.onFlashLoan"));
}
```
And, indeed, we help the solver by assuming that the only `_receiver` can only be **SymbolicAttacker**:
```solidity
function flashLoan(...)
{
    vm.assume(address(_receiver) == address(0xaaaa0006));
    ...
}

> ```

Reviewer: it feels a little dirty to add the vm.assume directly in `SelfiePool`. Who is making that call? could we add the assume at the callsite instead?

Author: This call make SymbolicAttacker symbolically. We can do it on SymbolicAttacker's side, but imagine all this logic. It'll look something like
```solidity
...
if (bytes4(symbolic_data) == "flashLoan(...)")
{
vm.assume( <logic to assume that _receiver is 0xaaaa0006>)
}
...
target.call(symbolic_data);
```

Also, imagine if such a assumption happens somewhere deeper. We can't do all this stuff conveniently in SymbolicAttacker (or we are doomed to craft parameters like they do in Echidna, before transaction start). 

So, it looks like the most readable way to help Halmos make right assumptions, little changes right in the target contracts

Reviewer: having that logic in `SymbolicAttacker` doesn't seem shocking to me, at least you know where the assumes are

    GovernanceAction memory actionToExecute = _actions[actionId];

    if (actionToExecute.proposedAt == 0) return false;

    uint64 timeDelta;
    unchecked {
        timeDelta = uint64(block.timestamp) - actionToExecute.proposedAt;
    }

    return actionToExecute.executedAt == 0 && timeDelta >= ACTION_DELAY_IN_SECONDS;
}
```
This revert is here because we never had any **action** registered. Since we could not symbolically enter here during 1 transaction, we assume that at least 2 are necessary: the first one registers the **action**, the second executes it. We also pay attention to the use of `block.timestamp`. This is a clear indication that some time should pass between transactions. We can't cover this code directly at this point, as we don't know the way to register an **action**. But we know for sure that one symbolic transaction will not be enough for us.

So our **SymbolicAttacker** becomes extended:
```solidity
function attack() public {
    execute_tx();
    uint256 warp = svm.createUint256("warp");
    vm.warp(block.timestamp + warp); // wait for symbolic time between transactions
    execute_tx();

> }

Reviewer: nice, makes sense

3. Simplify/change the invariant to make the engine's task easier.

Let's go through each of these points.
### Solid optimizations
The first thing I can think of here is to completely exclude `ERC20::permit` from symbolic function candidates, which is already starting to get annoying. I can't think of any scenario where we could apply it where `ERC20Votes::approve` is not applicable:
```solidity
/*
** This function has the same purpose as get_concrete_from_symbolic, 
** but applies optimizations and heuristics.
*/
function get_concrete_from_symbolic_optimized (address /*symbolic*/ addr) public view 
                                    returns (address ret, bytes memory data) 
{
    for (uint256 i = 0; i < addresses_list_size; i++) {
        if (addresses[i] == addr) {
            string memory name = names_by_addr[addresses[i]];
            ret = addresses[i];
            data = svm.createCalldata(name);
            bytes4 selector = svm.createBytes4("selector");
            vm.assume(selector == bytes4(data));
            // Not DamnValuableVotes permit
            vm.assume(selector != bytes4(keccak256("permit(address,address,uint256,uint256,uint8,bytes32,bytes32)")));

>             return (ret, data);

Reviewer: do you create a separate symbolic `selector` for counterexample readability? otherwise you should be able to do this directly:

`vm.assume(bytes4(data) != ERC20Permit.permit.selector);`

Author: Actually, I changed this logic in "backdoor". to have more convenient interface for excluding such functions. I think I should rewrite this logic here like I did in backdoor, no reason to leave it as is.

```solidity
function get_concrete_from_symbolic_optimized (...) 
{
    ...
    // Not DamnValuableVotes::delegateBySig
    vm.assume(selector != bytes4(keccak256("delegateBySig(address,uint256,uint256,uint8,bytes32,bytes32)")));
    ...
}
```
### Cut scenarios
Let's try to cut down the scenarios in which we symbolically enter the same function several times. Now we can't select the same function twice inside `get_concrete_from_symbolic` in the path. At the same time, the overall coverage of the code will not decrease, we will still go through all scenarios where these functions are entered once:
```solidity
contract GlobalStorage is Test, SymTest {
    ...
    mapping (uint256 => bytes4) used_selectors;
    uint256 used_selectors_size = 0;
    ...
    function get_concrete_from_symbolic_optimized (...) 
    {
        ...
        for (uint256 s = 0; s < used_selectors_size; s++) {
            vm.assume(selector != used_selectors[s]);
        }
        used_selectors[used_selectors_size] = selector;
        used_selectors_size++;
        ...
    }
    ...
}
```

> Another sacrifice of scenarios in exchange for avoiding duplication coverage can be achieved if you expand the number of symbolic transactions not in `attack()`, but in `onFlashLoan()` callback. This way we still process 2 symbolic transactions, but only if a **flashLoan** happened, which saves us a lot of resources:

Reviewer: ok this is clever but goes by very fast! I had to read it a couple times to process it properly

Author: OK, I'll try to rewrite this moment in more clear manner

```solidity
...
function execute_tx(string memory target_name) private {
        address target = svm.createAddress(target_name);
        bytes memory data;
        //Get some concrete target-name pair
        (target, data) = glob.get_concrete_from_symbolic_optimized(target);
        target.call(data);
    }

    function onFlashLoan(address initiator, address token,
                        uint256 amount, uint256 fee,
                        bytes calldata data
    ) external returns (bytes32) 
    {
        execute_tx("onFlashLoan_target1");
        uint256 warp = svm.createUint256("onFlashLoan_warp");
        vm.warp(block.timestamp + warp); // wait for symbolic time between transactions
        execute_tx("onFlashLoan_target2");

>         DamnValuableVotes(token).approve(address(msg.sender), 2**256 - 1); // unlimited approve for pool

Reviewer: ```suggestion
        DamnValuableVotes(token).approve(address(msg.sender), type(uint256).max); // unlimited approve for pool
```

        return (keccak256("ERC3156FlashBorrower.onFlashLoan"));
    }
...
function attack() public {

>     execute_tx();

Reviewer: missing the target name here

Author: True, will fix

    /*uint256 warp = svm.createUint256("warp");
    vm.warp(block.timestamp + warp); // wait for symbolic time between transactions
    execute_tx();*/
}
```

> Yes, we have moved the `vm.warp()` from `attack()` to `onFlashLoan()` right in the middle of the transaction. Because we can! It may be risky because it can cause **false positives**. We hope this will not happen. This is the price of optimization heuristics.

Reviewer: why is it necessary to move the warp to onFlashLoan? aren't we just trying to queue a governance action, not execute it?

Author: Actually, we are trying to break ANY invariant. Since we have logic with timestamp and we don't know the exact place where it should be used (and also we have only one symbolic transaction in `attack()`, we have moved second tx to the `onFlashLoan()` ) this is the only place where warp may be used. At this time we can't know do we need warp or not.

What do you think? Should it be rewritten in more clear manner?

Reviewer: but I don't see how adding the warp in `onFlashLoan()` is an optimization if it can create false positives. It's allowing more exploration, not less

### Invariants
Until now, we used only invariants that somehow followed from the initial conditions of the problem. I suggest this time to go a much more creative way and come up with any scenarios that seem unexpected, unnatural or buggy. Yes, let's do the work for the imaginary developers of these contracts and cover them with tests :D.

> Let's start with token **allowance**. It is unexpectedly, that as a result of the **attacker's** actions, the **pool's** or **governance's** allowance may somehow change:

Reviewer: ```suggestion
Let's start with token **allowance**. It is unexpected, that as a result of the **attacker's** actions, the **pool's** or **governance's** allowance may somehow change:
```

```solidity
function _isSolved() private view {
    ...
    // Check allowance changes
    address symbolicSpender = svm.createAddress("symbolicSpender");
    assert (token.allowance(address(pool), symbolicSpender) == 0);
    assert (token.allowance(address(governance), symbolicSpender) == 0);
    ...
}
```
in **simpleGovernance** contract `_votingToken` is not immutable:
```solidity
contract SimpleGovernance is ISimpleGovernance {
    ...
    DamnValuableVotes private _votingToken;
    ...
}
```
Perhaps there is a scenario in which we can somehow change it:
```solidity
function _isSolved() private view {
    ...
    //check if governance's _votingToken may be changed
    assert (governance._votingToken() == token);
    ...

> }

Reviewer: very nice, love this kind of ideas

    bytes memory data = svm.createBytes(1000, "preload_data");
    uint256 amount = svm.createUint256("preload_amount");
    pool.flashLoan(IERC3156FlashBorrower(address(this)), address(token), amount, data);
    is_preload = false;
}
...
```
We do not forget to remove assert for the constancy of `_actionCounter`, otherwise every path will be a counterexample:
```solidity
function _isSolved() private view {
    ...
    // Check number of registered actions
    //assert (governance._actionCounter() == 1);
}
```
Since we unlocked the `executeAction` function, let's start again with one symbolic transaction. Will see if that's enough.

Run:
```javascript
$ halmos --solver-timeout-assertion 0 --function check_selfie --loop 3
...
WARNING  Encountered symbolic memory offset: 320 + Concat(...)
...
```
The problem here is:
```solidity
function executeAction(uint256 actionId) external payable returns (bytes memory) {
...
GovernanceAction storage actionToExecute = _actions[actionId];
...
```
Obviously, we have only one action. Feel free to use `assume`:
```solidity
function executeAction(uint256 actionId) external payable returns (bytes memory) {
    vm.assume(actionId == 1);
    ...

> ```

Reviewer: I can't immediately see why that warning relates to the action lookup, weird 🤨 @daejunpark does it make sense to you?

Reviewer 2: good question. i guess it might be related to the copy of GovernanceAction struct into memory (which happens inside _canBeExecuted() called by executeAction().)

the GovernanceAction struct contains a `bytes data` field, and copying it may cause the symbolic memory offset error.

Also at this point we have recursion, as some function from **SymbolicAttacker** can be launched as a target.
We remove such a scenario:
```solidity

> vm.assume(actionToExecute.target != address(0xaaaa0006));

Reviewer: a comment would be useful here, is 0xaaaa0006 the address of SymbolicAttacker?

Author: OK, I'll write comment here. But this is the exact reason why I print all of these addresses in the start of article)

Reviewer: well yeah but the start of the article is far away at this point and I don't commit addresses to memory 🤪

    // initialize payload variables
    bytes memory _payload;
    address _receiver;
    // either create a payload of drainAllFunds funtion if selected
    if (_payloadNum == uint256(PayloadTypesInQueueAction.drainAllFunds)) {
        _payload = abi.encodeWithSignature(
            "drainAllFunds(address)",
            address(this)
        );
        _receiver = address(pool);
    }
    // or create a payload of transferFrom function if selected
    if (_payloadNum == uint256(PayloadTypesInQueueAction.transferFrom)) {
        // _transferAmountInPayload.push(_amountToTransfer);
        _payload = abi.encodeWithSignature(
            "transferFrom(address,address,uint256)",
            address(pool),
            address(this),
            _amountToTransfer
        );
        _receiver = address(token);
        // store amount to transfer
        payloads[_counter].transferAmount = _amountToTransfer;
    }
    // fill payload mapping by the variables created
    payloads[_counter].payload = _payload;
    payloads[_counter].receiver = _receiver;
    // increase payload counter (for the next iteration of new payload creation)
    ++payloadsPushedCounter;
}
```

> A bit like our Frankenstein from [Truster](https://github.com/igorganich/damn-vulnerable-defi-halmos/tree/master/test/truster#echidna), isn't it? Again, I draw your attention to the fact that only 2 functions that can be launched inside `executeAction` are considered here, and the function is already so huge. And the parameters of these functions are not abstract, but hardened. Against this background, the usability of Halmos is obvious.

Reviewer: ```suggestion
A bit like our Frankenstein from [Truster](https://github.com/igorganich/damn-vulnerable-defi-halmos/tree/master/test/truster#echidna), isn't it? Again, I draw your attention to the fact that only 2 functions that can be launched inside `executeAction` are considered here, and the function is already so huge. And the parameters of these functions are not abstract, but hard-coded. Against this background, the usability of Halmos is obvious.
```
